### PR TITLE
Pinning Notebooks on Notebooks view

### PR DESCRIPTION
**Pull request by MaddyDev**
Added pinning feature to the Notebooks viewlet which adds a new viewlet "Pinned Notebooks" to the existing Notebooks explorer view. PinnedNotebooks is only visible when users have pinned notebooks in their workspace. Pinned notebooks are specific to workspaces. (If user switches between different workspaces, pinned notebooks specific to that workspace are only shown.)
Following actions are supported:
1. Pin notebooks from Notebooks viewlet.
2. Remove pinned notebooks from PinnedNotebooks viewlet.
Note: Pinning is not applicable to ProvidedNotebooks. 

Removing pinned notebooks from Notebooks viewlet is currently not supported (looking into a potential vscode bug that's blocking me from implementing this).

pinningNotebooks

### CODE REVIEW

**Comment by MaddyDev** [comment]
@yualan Can you help me with proper messages to display. I have temporarily added messages that I thought fit.

File: extensions/notebook/src/book/bookTreeView.ts [hunk 1]
```diff
@@ -67,6 +74,17 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
...
 
+	private async initializePinnedNotebooks(): Promise<void> {
+		await Promise.all(getPinnedNotebooks().map(async (notebookPath) => {
+			try {
+				await this.createAndAddBookModel(notebookPath, true);
+			} catch {
+				// no-op, not all workspace folders are going to be valid books
+			}
+		}));
+		this._initializeDeferred.resolve();
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/book/bookTreeView.ts:85]: You should only be calling this once - you're already calling it in intialize.

Would suggest moving the calling of this function into initialize

File: extensions/notebook/src/book/bookTreeView.ts [hunk 2]
```diff
@@ -97,6 +115,17 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
 		}
 	}
 
+	async pinNotebook(bookTreeItem: BookTreeItem, unpin: boolean = false): Promise<void> {
+		let bookPathToUpdate = bookTreeItem.book?.contentPath;
+		if (bookPathToUpdate) {
+			let pinStatusChanged = unpin ? this.bookPinManager.unpinNotebook(bookTreeItem) : this.bookPinManager.pinNotebook(bookTreeItem);
+			if (pinStatusChanged) {
+				bookTreeItem.contextValue = unpin ? 'savedNotebook' : 'pinnedNotebook';
+				vscode.window.showInformationMessage(unpin ? loc.msgBookUnpinned(bookTreeItem.title) : loc.msgBookPinned(bookTreeItem.title));
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/book/bookTreeView.ts:124]: I think the message is unnecessary here - that's spammy and it seems pretty clear to me already what's happening.

File: extensions/notebook/src/extension.ts [hunk 3]
```diff
@@ -39,6 +39,15 @@ export async function activate(extensionContext: vscode.ExtensionContext): Promi
 	extensionContext.subscriptions.push(vscode.commands.registerCommand('notebook.command.closeBook', (book: any) => bookTreeViewProvider.closeBook(book)));
 	extensionContext.subscriptions.push(vscode.commands.registerCommand('notebook.command.closeNotebook', (book: any) => bookTreeViewProvider.closeBook(book)));
 	extensionContext.subscriptions.push(vscode.commands.registerCommand('notebook.command.openNotebookFolder', (folderPath?: string, urlToOpen?: string, showPreview?: boolean,) => bookTreeViewProvider.openNotebookFolder(folderPath, urlToOpen, showPreview)));
+	extensionContext.subscriptions.push(vscode.commands.registerCommand('notebook.command.pinNotebook', async (book: any) => {
+		await bookTreeViewProvider.pinNotebook(book);
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/extension.ts:43]: Would suggest just making separate pin/unpin functions...unpinning a notebook by calling pinNotebook is weird.

File: extensions/notebook/src/extension.ts [hunk 4]
```diff
@@ -39,6 +39,15 @@ export async function activate(extensionContext: vscode.ExtensionContext): Promi
...
+		await pinnedBookTreeViewProvider.addRemoveNotebookFromPinned(book);
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/extension.ts:44]: Same here - just have separate add/remove functions

**Comment by Charles-Gagnon** [comment]
A couple follow-up questions though : 

- What happens if the book containing a pinned notebook is closed from the Viewlet?
- What happens if the notebook is deleted from the disk?
- What's the issue here? `Removing pinned notebooks from Notebooks viewlet is currently not supported`

File: extensions/notebook/resources/light/pin.svg [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (Charles-Gagnon) [inline extensions/notebook/resources/light/pin.svg:1]: You don't need the pin images anymore, right?

File: extensions/notebook/src/book/bookPinManager.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+	isNotebookPinned(notebookPath: string): boolean {
+		if (getPinnedNotebooks().findIndex(x => x === notebookPath) > -1) {
+			return true;
+		}
+		return false;
+	}
+
+	pinNotebook(notebook: BookTreeItem): boolean {
+		return this.isNotebookPinned(notebook.book.contentPath) ? false : this.updatePinnedBooks(notebook, PinBookOperation.Pin);
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/book/bookPinManager.ts:43]: So this returns false if the notebook isn't actually pinned but unpinNotebook always returns updatePinnedBooks directly? It looks like the logic for updatePinnedBooks will return true/false as appropriate so you should be able to get rid of that check here right?

File: extensions/notebook/src/book/bookTreeView.ts [hunk 7]
```diff
@@ -97,6 +111,26 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
 		}
 	}
 
+	async pinNotebook(bookTreeItem: BookTreeItem, unpin: boolean = false): Promise<void> {
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/book/bookTreeView.ts:114]: You no longer user unpin

File: extensions/notebook/src/common/utils.ts [hunk 8]
```diff
@@ -323,3 +323,19 @@ export async function getRandomToken(size: number = 24): Promise<string> {
 		});
 	});
 }
+
+export function isBookItemPinned(notebookPath: string): boolean {
+	let config: vscode.WorkspaceConfiguration = vscode.workspace.getConfiguration(notebookConfigKey);
+	let pinnedNotebooks: string[] = config.get(pinnedBooksConfigKey);
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/common/utils.ts:329]: nit: call getPinnedNotebooks here to consolidate the logic

**Comment by MaddyDev** [comment]
> A couple follow-up questions though :
> 
> * What happens if the book containing a pinned notebook is closed from the Viewlet?
> * What happens if the notebook is deleted from the disk?
> * What's the issue here? `Removing pinned notebooks from Notebooks viewlet is currently not supported`

> * Even if the book is closed as long as the user works in the workspace, pinned notebooks in the workspace are shown.
> * I need to test this scenario but am expecting it would fail saying the path cannot be found. Will test and address this gracefully.
> * Once a notebook is pinned, am updating the tree item context value to show unpin icon instead of pin. For some reason UI isn't reflecting the updated context.

**Pull request merged**